internal/configs/version2/nginx.virtualserver.tmpl

```diff
@@ -8,8 +8,10 @@ upstream {{ $u.Name }} {
     server {{ $s.Address }} max_fails={{ $s.MaxFails }} fail_timeout={{ $s.FailTimeout }};
     {{ end }}
 
-    {{ if $.Keepalive }}
-    keepalive {{ $.Keepalive }};
+    {{ if $u.Keepalive }}
+    keepalive {{ $u.Keepalive }};
+    {{ else }}
+    {{ if $.Keepalive }}keepalive {{ $.Keepalive }};{{ end }}
```

Comment:
There is no need to have `$.Keepalive` / `version2.VirtualServerConfig.Keepalive` now, that was done because it wasn't possible to specify a keepalive for each upstream, but I think that can be replaced now for the `Upstream.Keepalive` now that you have implemented it 🤔 

PS: I think this directive also needs to add the `proxy_set_header Connection "";` in the location blocks of the upstream (ref https://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive) you can also have a look at the template for the Ingress Resource to see this: https://github.com/nginxinc/kubernetes-ingress/blob/master/internal/configs/version1/nginx.ingress.tmpl#L127